### PR TITLE
Proxy read ids from redis

### PR DESCRIPTION
**Pull request by khanova**
## Problem

Proxy doesn't know about existing endpoints.

## Summary of changes

* Added caching of all available endpoints. 
* On the high load, use it before going to cplane.
* Report metrics for the outcome.
* For rate limiter and credentials caching don't distinguish between `-pooled` and not

TODOs:
* Make metrics more meaningful
* Consider integrating it with the endpoint rate limiter
* Test it together with cplane in preview

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: proxy/src/context.rs [hunk 1]
```diff
@@ -49,6 +51,8 @@ pub struct RequestMonitoring {
     // This sender is here to keep the request monitoring channel open while requests are taking place.
     sender: Option<mpsc::UnboundedSender<RequestData>>,
     pub latency_timer: LatencyTimer,
+    // Whether proxy decided that it's not a valid endpoint end rejected it before going to cplane.
+    rejected: bool,
```
Comment (conradludgate) [inline proxy/src/context.rs:55]: Should this instead be an error kind?

Comment (conradludgate) [reply proxy/src/context.rs:55]: Oh, I see you are using it for metrics.

File: proxy/src/console/provider.rs [hunk 2]
```diff
@@ -538,12 +545,9 @@ impl ApiLocks {
         })
     }
 
-    pub async fn garbage_collect_worker(&self, epoch: std::time::Duration) {
-        if self.permits == 0 {
-            return;
-        }
```
Comment (conradludgate) [inline proxy/src/console/provider.rs:544]: Why did you remove this check?

File: proxy/src/cache/endpoints.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    ProjectCreated,
+}
+
+pub struct EndpointsCache {
+    config: EndpointCacheConfig,
+    endpoints: DashSet<EndpointIdInt>,
+    branches: DashSet<BranchIdInt>,
+    projects: DashSet<ProjectIdInt>,
+    ready: AtomicBool,
+    limiter: Arc<Mutex<GlobalRateLimiter>>,
```
Comment (conradludgate) [inline proxy/src/cache/endpoints.rs:41]: Can we use a sync lock here? I'm also going to review whether this lock will add too much contention since under high load this will be constantly locked and unlocked.

Comment (conradludgate) [reply proxy/src/cache/endpoints.rs:41]: 93ns... that's 10 million a second, this is benchmarking with a lock too... ok this is fine :)

File: proxy/src/console/provider.rs (hunk 2 above)
Comment (khanova) [reply proxy/src/console/provider.rs:542]: I guess, it's never the case that there are 0 permits and it shouldn't hurt  much.

I changed return type to `Infallible` to put it in the maintenance tasks.

File: proxy/src/cache/endpoints.rs (hunk 3 above)
Comment (khanova) [reply proxy/src/cache/endpoints.rs:41]: So, basically, keeping mutex is fine, right?
